source/common/thread_local/thread_local_impl.cc

```diff
@@ -26,79 +26,73 @@ SlotPtr InstanceImpl::allocateSlot() {
   ASSERT(!shutdown_);
 
   if (free_slot_indexes_.empty()) {
-    SlotImplPtr slot(new SlotImpl(*this, slots_.size()));
-    auto wrapper = std::make_unique<Bookkeeper>(*this, std::move(slot));
-    slots_.push_back(wrapper->slot_.get());
-    return wrapper;
+    SlotPtr slot(new SlotImpl(*this, slots_.size()));
+    slots_.push_back(slot.get());
+    return slot;
   }
   const uint32_t idx = free_slot_indexes_.front();
   free_slot_indexes_.pop_front();
   ASSERT(idx < slots_.size());
-  SlotImplPtr slot(new SlotImpl(*this, idx));
+  SlotPtr slot(new SlotImpl(*this, idx));
   slots_[idx] = slot.get();
-  return std::make_unique<Bookkeeper>(*this, std::move(slot));
+  return slot;
 }
 
-bool InstanceImpl::SlotImpl::currentThreadRegistered() {
-  return thread_local_data_.data_.size() > index_;
-}
+InstanceImpl::SlotImpl::SlotImpl(InstanceImpl& parent, uint32_t index)
+    : parent_(parent), index_(index), still_alive_guard_(std::make_shared<bool>(true)) {}
 
-void InstanceImpl::SlotImpl::runOnAllThreads(const UpdateCb& cb) {
-  parent_.runOnAllThreads([this, cb]() { setThreadLocal(index_, cb(get())); });
+Event::PostCb InstanceImpl::SlotImpl::wrapCallback(Event::PostCb cb) {
+  // See the header file comments for still_alive_guard_ for the purpose of this capture.
+  return [still_alive_guard = std::weak_ptr<bool>(still_alive_guard_), cb] {
+    if (!still_alive_guard.expired()) {
```

Comment:
If I understand correctly, the logic here is that if it is !expired() at the start of this operation, we don't care if it gets destructed concurrently with cb() running, because the destruction will get post()'d to this thread and thus won't actually destruct the thread-local state on this thread until cb() is done. Is that correct? Can you add more comments here?

Reply:
Yes, correct, will add more comments.